New version of the script_score term stats helpers.

afoucret opened the pull request:
A new version of the script_score term stats helpers.

## Overview

This PR exposes term statistics for the child query through the `_termStatistics` variable (instead of having parameters to specify the field and query as parameters of the helper).

As an example, the following query: 

```
{
  "query": {
    "script_score": {
      "query": { "match": { "title": "star wars" } },
      "script": {
        "source": "return _termStatistics.docFreq().getSum()"
      }
    }
  }
}
```

will read document frequency for terms `star` and `wars` in the title field and return their sum.

The following method are available in the `_termStatistics` object:


long uniqueTermsCount() | Number of unique terms in the query
-- | --
long matchedTermsCount() | Number of terms of the query matched in the current document
DoubleSummaryStatistics docFreq() | Statistics for the document frequency of the terms (number of documents in the index with the term).
DoubleSummaryStatistics totalTermFreq() | Statistics for the total term frequency of the terms (number of occurrence for a term in the index with the term).
DoubleSummaryStatistics termFreq() | Statistics for the term frequency of the terms (number of occurrence of the term in the current document)
DoubleSummaryStatistics termPositions() | Statistics for the term position of the terms in the current document

Commit:
New version of the script_score term stats helpers.

Commit:
Code refactoring so we are using _termStatistics reserved variable.

ioanatia commented on server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java:
```diff
@@ -86,9 +89,18 @@ public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float bo
             return subQuery.createWeight(searcher, scoreMode, boost);
```
when is this branch executed? and do we still need the `needsTermStatistics` check here and do `this.visit(QueryVisitor.termCollector(terms));`?

ioanatia commented on server/src/main/java/org/elasticsearch/script/ScoreScript.java:
```diff
@@ -92,11 +97,15 @@ public ScoreScript(Map<String, Object> params, SearchLookup searchLookup, DocRea
             this.params = null;
             ;
```
completely unrelated to your changes, but I guess this line is not needed, no?

afoucret commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.score.txt:
```diff
@@ -13,6 +13,21 @@ class org.elasticsearch.script.ScoreScript @no_import {
 class org.elasticsearch.script.ScoreScript$Factory @no_import {
 }
 
+class org.apache.lucene.index.Term {
+    String field()
+    String text()
+}
```
TODO : remove this cause we do not want to expose it

afoucret commented on modules/lang-painless/src/main/resources/org/elasticsearch/painless/org.elasticsearch.script.score.txt:
```diff
@@ -13,6 +13,21 @@ class org.elasticsearch.script.ScoreScript @no_import {
+}
+
+class org.elasticsearch.script.TermStatsReader {
+    Set terms()
```
TODO : remove this cause we do not want to expose it

jimczi reviewed (commented):
The approach looks good to me, it is isolated and script_score queries that don't need the term statistics won't be impacted so +1.

@rjernst you had some concerns regarding the re-introduction of term stats in script, would this approach alleviates those?

afoucret committed:
Merge branch 'elastic:main' into term-stats-script-score-v2

Commit:
Fix test compilation error.

Commit:
Fixing failing tests.

Commit:
Code refactoring.

Commit:
Remove useless changes.

Commit:
Additional refactoring.

Commit:
Adding tests.

Commit:
Update tests.

afoucret commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/190_term_statistics_script_score.yml:
```diff
@@ -0,0 +1,612 @@
+setup:
```
ℹ️ YAML rest test for many use case.

afoucret committed:
Merge branch 'elastic:main' into term-stats-script-score-v2

Commit:
Update comments

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

afoucret commented:
@elastic/es-core-infra would be great if someone could review this PR

Commit:
Lint

Commit:
Simplify if clause.

afoucret commented on server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java:
```diff
@@ -86,9 +90,18 @@ public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float bo
         boolean needsScore = scriptBuilder.needs_score();
-        ScoreMode subQueryScoreMode = needsScore ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+        boolean needsTermStatistics = scriptBuilder.needs_termStatistics();
+
```
ℹ️ Required so termStats are collected during DFS when `needs_termStatistics` is true.

afoucret commented on server/src/main/java/org/elasticsearch/script/ScoreScript.java:
```diff
@@ -189,14 +192,38 @@ public void _setIndexName(String indexName) {
+        /**
+         * Return {@code true} if the script needs {@code _termStatistics} calculated, or {@code false} otherwise.
+         */
+        default boolean needs_termStatistics() {
```
ℹ️ This methods is automatically implemented by the painless script factory and return `true` if the `_termStatistics` variable is used in the script.

afoucret replied:
I wanted to avoid reimplementing every time we have a anonymous class inheriting from ScriptScore (there is a lot of them in the tests). But I changed my mind since it is probably weird to have a default implementation for it.

Commit:
Refactoring to make the code easier to read.

Commit:
Remove useless unrelated changes.

rjernst reviewed (commented):
@afoucret My concerns of performance hits from reintroducing term stats are mostly allayed here. This PR is well put together, thank you!

Most of my comments are small. One larger concern is over the caching strategy. I think we should avoid maps. Although using HashMap has constant time lookup, we still must compute the hash N * T (number of documents scored, number of terms in query) times. Unlike the previous term stats API in scripts here we do not give access to per term stats (which is a great simplification!). But then we don't need the associative lookup, just the ability to iterate over the terms enums/states.

Additionally, I think we should have unit tests for each of the methods on ScriptTermStats (when retrieved from a ScoreScript), making sure to hit all edge cases handled in these methods.

Finally, I have a very minor bikeshed about the naming. I see you mimicked the underscore prefix of score. Note that `_score` is somewhat a historical name that we've been stuck with. And since we don't restrict variable names from users, underscore isn't actually a protected namespace (though maybe it should be!). One concrete feedback we've received in the past with painless is long names. So I wonder if it should just be `termStats` (or `_termStats`).

Commit:
Merge branch 'main' of https://github.com/elastic/elasticsearch into term-stats-script-score-v2

Commit:
Apply suggestion from review.

afoucret commented on server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:
```diff
@@ -0,0 +1,222 @@
+        for (Term term : terms) {
+            try {
+                PostingsEnum postingsEnum = termStatsReader.postings(term);
+                int docId = docIdSupplier.get();
```
Done.

Commit:
Code refactoring:
- Create a statistic accumulator that can be reused
- Change the caching strategy for term states and enum

Commit:
Minor refactoring.

benwtrent commented on server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:
```diff
@@ -0,0 +1,240 @@
+            try {
+                TermStates termStates = TermStates.build(searcher, term, true);
+
+                // Using the searcher to get term statistics. If search_type is dfs, this is how the stats from the DFS phase are read.
+                return searcher.termStatistics(term, termStates.docFreq(), termStates.totalTermFreq());
+            } catch (IllegalArgumentException e) {
+                return null;
+            }
```
When does a IAE throw here? It seems weird to just ignore it.

benwtrent reviewed (commented):
Can we have a test similar to `NeedsScoreTests`?

Overall, this looks good. I do wonder about the default `leafFactor` impl as Ryan does.

Commit:
Refactoring: remove useless ScriptTermStats.Reader

Commit:
Rename _termStatistics into _termStats

afoucret commented on server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:
```diff
@@ -0,0 +1,222 @@
+     * @return statistics on docFreq for the terms of the query.
+     */
+    public DoubleSummaryStatistics docFreq() {
+        DoubleSummaryStatistics docFreqStatistics = new DoubleSummaryStatistics();
```
I introduced a new `StatsAccumulator` class that does more or less the same than `DoubleSummaryStatistics` but that can be reset.

afoucret commented on server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:
```diff
@@ -0,0 +1,222 @@
+            this.leafReaderContext = leafReaderContext;
+        }
+
+        PostingsEnum postings(Term term) {
```
The code has been refactored and should looks like what you described. Is it better right now?

Commit:
Remove default implementation for need_termStats.

Commit:
Rely on CachedSupplier instead of volatile properties.

Commit:
StatsAccumulator unit tests.

Commit:
Add termStats unit tests to ScriptScoreQuery.

Commit:
Test the need_termStats is implemented correctly for score script.

afoucret commented on server/src/test/java/org/elasticsearch/search/query/ScriptScoreQueryTests.java:
```diff
@@ -177,6 +177,11 @@ public boolean needs_score() {
 
+            @Override
+            public boolean needs_termStatistics() {
+                return randomBoolean();
```
I did implement two tests cases to check if the termStats is available only if `needs_termStats` is true

afoucret commented on server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:
```diff
@@ -0,0 +1,222 @@
+                try {
+                    TermStates termStates = termStates(term);
+
+                    if (termStates == null || termStates.docFreq() == 0) {
```
Indeed, it is never the case. I removed the conditional for it.

Commit:
Fix testNeedsTermStats (update to the right script context).

Commit:
Improved unit testing.

afoucret commented:
@rjernst @benwtrent, I think all the comments from your reviews have been addressed through the refactoring and adding a lot of tests. I would be glad to read your review again.

Commit:
Remove call to forbidden API,

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:
```diff
@@ -0,0 +1,233 @@
+ * Access the term statistics of the children query of a script_score query.
+ */
+public class ScriptTermStats {
+    private final Supplier<Integer> docIdSupplier;
```
This could be `IntSupplier`

rjernst reviewed (commented):
This looks really close. I have a few more thoughts.

Commit:
Fix test regression

Commit:
Rename StatsAccumulator into StatsSummary

Commit:
Change docIdSupplier type to IntSupplier

afoucret commented on server/src/main/java/org/elasticsearch/script/StatsAccumulator.java:
```diff
@@ -0,0 +1,123 @@
+ * <p>This class provides statistics such as count, sum, minimum, maximum, and arithmetic mean
+ * of the recorded values.
+ */
+public class StatsAccumulator implements DoubleConsumer {
```
Renamed into `SfatsSummary` per your suggestion.

Commit:
Introduce a new cluster features for term stats.

afoucret commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/190_term_statistics_script_score.yml:
```diff
@@ -0,0 +1,612 @@
+setup:
+  - requires:
+      cluster_features: ["gte_v8.16.0"]
```
I introduced a new feature instead.

afoucret commented on server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:
```diff
@@ -0,0 +1,233 @@
+            for (int i = 0; i < termContexts.length; i++) {
+                try {
+                    totalTermFreqStats.accept(termStatistics(terms[i], termContexts[i]).totalTermFreq());
+                } catch (IllegalArgumentException e) {
```
`termStatistics` is using the `Searcher::termStatistics` method:

```
private TermStatistics termStatistics(Term term, TermStates termStates) throws IOException {
    return searcher.termStatistics(term, termStates.docFreq(), termStates.totalTermFreq());
}
When a term is not present in the current shards, this is equivalent to:

searcher.termStatistics(term, 0, 0);
which uses aggregated Document Frequency Statistics (DFS) to retrieve term statistics:

if (aggregatedDfs == null) {
    // Either the DFS phase is executing, or the search type does not include a DFS phase.
    return super.termStatistics(term, docFreq, totalTermFreq);
}

TermStatistics termStatistics = aggregatedDfs.termStatistics().get(term);
if (termStatistics == null) {
    // No stats available for this term - it could be from a "must_not" clause or similar, 
    // preventing term extraction from the query.
    return super.termStatistics(term, docFreq, totalTermFreq);
}

return termStatistics;
```

when a term is missing it will try to create a `TermStatistics` object with a docFreq  equal to 0, causing the IAE to be raised.

Because we want to leverage the stats from dfs for more reliable results in the distributed use case, we have to catch this edge case since there is no way to predict when the IAE will be thrown.

rjernst replied:
This seems fragile. There could be other cases (for example, an actual bad argument, like reversing term freq and total term freq) for which an IAE is thrown. I think we should find a way to avoid calling termStatistics. For example, you could have a new method on ContextIndexSearcher, to find the totalTermFreq for a term (which takes a fallback). So something like:

```
if (searcher instanceof ContextIndexSearcher dfsSearcher) {
    totaltermFreqStats.accept(dfsSearcher.getTotalTermFreq(term, totalTermFreq));
} else {
    totaltermFreqStats.accept(totalTermFreq);
}
```

afoucret replied:
I like your proposal and implemented something very similar in the last version I just pushed.

afoucret commented on server/src/main/java/org/elasticsearch/script/ScoreScript.java:
```diff
@@ -189,14 +191,35 @@ public void _setIndexName(String indexName) {
+     */
+    public ScriptTermStats get_termStats() {
+        if (termStats == null) {
+            throw new IllegalArgumentException("_termStats is not available");
```
`_termStats` is currently implemented in the `script_score` query only.

While we want to implement it in some other place (function_score), it is not yet the case (should be easy to do once this PR will be merged). The exception makes it clear to the user that `_termStats` is not available in that case.

afoucret replied:
I did update it to an assertion per your recommendation.

afoucret commented on test/framework/src/main/java/org/elasticsearch/script/MockScriptEngine.java:
```diff
@@ -739,6 +739,11 @@ public boolean needs_score() {
 
+                @Override
+                public boolean needs_termStats() {
+                    return false;
```
I prefer to keep false since we would have to implement a mock for ScriptTermStats in that case.
BTW, there is specific tests for the termStats use case.

afoucret commented:
@rjernst I addressed most of your comment and feedback. It would be great if you have the time for another review (🤞 and maybe an approve this PR!!!!)

Commit:
Reafactor the way we are using the searcher.

Commit:
lint

Commit:
Use assert instead of exception in get_termStats

rjernst reviewed (approved):
Thanks for all the iterations. LGTM.

Commit:
lint

Commit:
Fix error in a test.

afoucret committed:
Merge branch 'elastic:main' into term-stats-script-score-v2
